src/style-spec/expression/definitions/within.js

```diff
@@ -0,0 +1,273 @@
+    }
+    return false;
+}
+
+function lineStringWithinPolygon(line, polygon) {
+    // First, check if geometry points of line segments are all inside polygon
+    for (let i = 0; i < line.length; ++i) {
+        if (!pointWithinPolygon(line[i], polygon)) {
+            return false;
+        }
+    }
+
+    // Second, check if there is line segment intersecting polygon edge
+    for (let i = 0; i < line.length - 1; ++i) {
+        if (lineIntersectPolygon(line[i], line[i + 1], polygon)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+function lineStringWithinPolygons(line, polygons) {
+    if (polygons.type === 'Polygon') {
+        return lineStringWithinPolygon(line, polygons.coordinates);
+    }
+    for (let i = 0; i < polygons.coordinates.length; i++) {
+        if (!lineStringWithinPolygon(line, polygons.coordinates[i])) return false;
+    }
+    return true;
+}
+
+function pointsWithinPolygons(ctx: EvaluationContext, polygonGeometry: GeoJSONPolygons, polyBBox: BBox) {
+    const pointBBox = [Infinity, Infinity, -Infinity, -Infinity];
+    const lngLatPoints = [];
+    for (const points of ctx.geometry()) {
+        for (const point of points) {
+            const p = getLngLatPoint(point, ctx.canonicalID());
+            lngLatPoints.push(p);
+            updateBBox(pointBBox, p);
+        }
+    }
+    if (!boxWithinBox(pointBBox, polyBBox)) return false;
+    for (let i = 0; i < lngLatPoints.length; ++i) {
+        if (!pointWithinPolygons(lngLatPoints[i], polygonGeometry)) return false;
+    }
+    return true;
+}
+
+function linesWithinPolygons(ctx: EvaluationContext, polygonGeometry: GeoJSONPolygons, polyBBox: BBox) {
+    const lineBBox = [Infinity, Infinity, -Infinity, -Infinity];
+    const lineCoords = [];
+    for (const line of ctx.geometry()) {
+        const lineCoord = getLngLatPoints(line, ctx.canonicalID());
+        lineCoords.push(lineCoord);
+        calcBBox(lineBBox, lineCoord, 'LineString');
+    }
+    if (!boxWithinBox(lineBBox, polyBBox)) return false;
+    for (let i = 0; i < lineCoords.length; ++i) {
+        if (!lineStringWithinPolygons(lineCoords[i], polygonGeometry)) return false;
+    }
+    return true;
+}
+
+class Within implements Expression {
+    type: Type;
+    geojson: GeoJSON
+    geometries: GeoJSONPolygons;
+    polyBBox: BBox;
+
+    constructor(geojson: GeoJSON, geometries: GeoJSONPolygons) {
+        this.type = BooleanType;
+        this.geojson = geojson;
+        this.geometries = geometries;
+        this.polyBBox = [Infinity, Infinity, -Infinity, -Infinity];
+        calcBBox(this.polyBBox, this.geometries.coordinates, this.geometries.type);
+    }
+
+    static parse(args: $ReadOnlyArray<mixed>, context: ParsingContext) {
+        if (args.length !== 2)
+            return context.error(`'within' expression requires exactly one argument, but found ${args.length - 1} instead.`);
+        if (isValue(args[1])) {
+            const geojson = (args[1]: Object);
```

Comment:
A nested expression would be fine with NSExpression on iOS/macOS; not sure about Android. I asked about this too in https://github.com/mapbox/mapbox-gl-js/pull/9352#discussion_r385956784 because I wasn’t sure whether to specifically prohibit a nested expression and only allow a literal shape.

Reply:
From my understanding, by applying the geometry with extra expression does the same thing. 
If we want to use `within` together with `get`, it needs to be format like 
```
[`within`, [`get`, `polygon`, `geojsonObj`]]
```
We any way need to pass the geojsonObj.  

Android SDK is trying to add binding for within expression. They are passing the argument as a map object, please check here https://github.com/mapbox/mapbox-gl-native-android/pull/198. 

I think one thing need to be updated is inside MGLConversion.h, currently `toGeoJSON()` is not implemented.

Reply:
```
[`within`, [`get`, `polygon`, `geojsonObj`]]
```
 or 
```
[`within`, [`literal`, `geojsonObj`]]
```
Another reason that it is better not to use a nested expression (like above) is, it will require evaluation of the inner expression for getting the geojsonObj every time when invoking `wtihin` evaluation, and it can not ensure if the geojsonObj provides valid geometry or not. If the geometry is invalid, it won't issue error when parsing the expression, only until evaluation stage, but that would be too late.

On the other hand, if the features are very dense or the geojsonObj contains pretty big amounts of points for the polygon, the process would be really inefficient. 

However, I do find something that `within` expression would break the style spec :
> 

> Expressions are represented as JSON arrays. The first element of an expression array is a string naming the expression operator, e.g. "*" or "case". Elements that follow (if any) are the arguments to the expression. Each argument is either a literal value (**a string, number, boolean, or null)**, or another expression array.

Style spec emphasized that valid argument could only be` a string, number, boolean, or null`,  but `within` requires  `object` argument.  @asheemmamoowala @mourner @mapbox/gl-js

Reply:
> Style spec emphasized that valid argument could only be` a string, number, boolean, or null`,  but `within` requires  `object` argument.

I think it was only a matter of time before array would have to start holding objects anyways. We kind of skirted around the issue when implementing `collator` objects, because we limited their use to a specific argument of a specific set of expression operators.

Reply:
> Another reason that it is better not to use a nested expression (like above) is, it will require evaluation of the inner expression for getting the geojsonObj every time when invoking `wtihin` evaluation, and it can not ensure if the geojsonObj provides valid geometry or not. If the geometry is invalid, it won't issue error when parsing the expression, only until evaluation stage, but that would be too late.

I think we’re on the same page. I was only referring to supporting something like `["within", ["match", …, geoJSONObj]]`. I wasn’t referring to `literal` expressions per se; sorry for being unclear with that terminology.